[FLINK-7337] [table] Refactor internal handling of time indicator attributes.

fhueske opened the pull request:
## What is the purpose of the change

Handle time indicators attributes as physical fields in `Row`.
- support for multiple event-time time attributes (required for proper join handling)
- no distinction of logical and physical schema

## Brief change log

- Expand phyiscal Row schema for time indicators.
- Refactor computation of logical schema of tables to import.
- `StreamRecord` timestamps are moved into `Row` during initial conversion
- Refactor operators to use time attribute in Row instead of StreamRecord timestamp.
- timestamps are copied into `StreamRecord` when `Table` is converted into `DataStream`.
- Drive-by fix for NPE in `generateInputFieldUnboxing`

## Verifying this change

- Changes are mostly internal. 
- No new features or public APIs have been added. 
- All existing tests pass.
- Tests have been added to verify that time indicators are copied when `Table` is converted into `DataStream`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **yes**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **na**

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/CRowInputTupleOutputMapRunner.scala:
```diff
@@ -90,3 +95,35 @@ class CRowInputScalaTupleOutputMapRunner(
+  * Wraps a ProcessFunction and sets a Timestamp field of a CRow as
+  * [[org.apache.flink.streaming.runtime.streamrecord.StreamRecord]] timestamp.
+  */
+class WrappingTimestampSetterProcessFunction[OUT](
```
Move this to separate file or rename file.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/CRowOutputProcessRunner.scala:
```diff
@@ -18,42 +18,54 @@
+  private var cRowWrapper: CRowWrappingCollector = _
 
   override def open(parameters: Configuration): Unit = {
     LOG.debug(s"Compiling MapFunction: $name \n\n Code:\n$code")
```
ProcessFunction

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/CRowOutputProcessRunner.scala:
```diff
@@ -18,42 +18,54 @@
   override def open(parameters: Configuration): Unit = {
     LOG.debug(s"Compiling MapFunction: $name \n\n Code:\n$code")
     val clazz = compile(getRuntimeContext.getUserCodeClassLoader, name, code)
     LOG.debug("Instantiating MapFunction.")
```
ProcessFunction

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/CRowOutputProcessRunner.scala:
```diff
@@ -18,42 +18,54 @@
+      out: Collector[CRow]): Unit = {
+
+    // remove timestamp from stream record
+    val tc = out.asInstanceOf[TimestampedCollector[_]]
```
Do we need this change? It is not executed if `org.apache.flink.table.plan.nodes.CommonScan#needsConversion` returns false.

fhueske replied:
It is not strictly required but reduces the serialization overhead by one Long value.
I added this to most functions that introduce a timestamp (ProcessFunction) but would also be OK to remove it.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1136,25 +1142,31 @@ object AggregateUtil {
-    val propPos = properties.foldRight((None: Option[Int], None: Option[Int], 0)) {
+    val propPos = properties.foldRight(
+      (None: Option[Int], None: Option[Int], None: Option[Int], 0)) {
       (p, x) => p match {
```
Could you name `x` to (a, b, c)? It is hard to read things like `(x._1, Some(x._4), x._3, x._4 - 1)`

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedRangeOver.scala:
```diff
@@ -123,6 +124,8 @@ class ProcTimeBoundedRangeOver(
       return
     }
 
+    out.asInstanceOf[TimestampedCollector[_]].eraseTimestamp()
```
Do we really need these erasing steps in every function? Shouldn't a rowtime operator overwrite it anyway?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala:
```diff
@@ -172,45 +172,20 @@ class FlinkTypeFactory(typeSystem: RelDataTypeSystem) extends JavaTypeFactoryImp
-    *                 specified fields, it shifts all following fields.
     * @return a struct type with the input fieldNames, input fieldTypes, and system fields
     */
   def buildLogicalRowType(
```
I think it's still fine. It creates a row type for the logical (Calcite) plan. 
But I'm fine changing the name if you have a better one.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala:
```diff
@@ -245,14 +244,18 @@ abstract class CodeGenerator(
       case Some(ti) =>
-        input2Mapping.map { idx =>
-          generateInputAccess(ti, input2Term, idx)
+        input2Mapping.map {
```
No, should not be used. I'll remove them.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamGroupWindowAggregate.scala:
```diff
@@ -148,6 +142,20 @@ class DataStreamGroupWindowAggregate(
+    val timestampedInput = if (isRowtimeAttribute(window.timeAttribute)) {
+      // copy the window rowtime attribute into the StreamRecord timestamp field
+      val timeAttribute = window.timeAttribute.asInstanceOf[ResolvedFieldReference].name
+      val timeIdx = inputSchema.logicalFieldNames.indexOf(timeAttribute)
```
Yes, in fact we can only do it like this because logical and physical type are the same. Otherwise, we would need to adjust the index.

I'll clean up `RowSchema` a bit more to remove the distinction of logical and physical types.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/RelTimeIndicatorConverter.scala:
```diff
@@ -415,7 +406,13 @@ class RexTimeIndicatorMaterializer(
         updatedCall.getOperands.map { o =>
           if (isTimeIndicatorType(o.getType)) {
-            rexBuilder.makeCall(TimeMaterializationSqlFunction, o)
+            if (isRowtimeIndicatorType(o.getType)) {
```
Not sure. The duplicated code is spread over 2 classes and differently parameterized. I tried to add a companion object to `RexTimeIndicatorMaterializer` and the method but this did not really improve the code, IMO.

fhueske commented:
Thanks for the review @twalthr. I addressed your comments and updated the PR.

I think it would be very good to handle the timestamps internally as longs. The change seems to be a bit more involved because we need to touch the serialization logic and various type conversion and code generation issues. I'd rather do this as a follow up to this PR. What do you think?

Fabian

twalthr commented:
I'm fine with a followup issue. +1 to merge this.

twalthr commented:
@wuchong @sunjincheng121 @shaoxuan-wang do you also want to take a look at it? Otherwise I would merge this and work on the followup issue for more efficiency.

wuchong commented:
@twalthr  I'm working on other issues before Thursday, so I would like to have a look at it at Thursday (Beijing). But if you are hurry, I'm fine to merge this first.

twalthr commented:
@wuchong No problem. Thursday is fine.

twalthr commented:
@wuchong could you take a look at #4532 as well? It is based on this PR.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/StreamTableSourceScan.scala:
```diff
@@ -46,29 +47,29 @@ class StreamTableSourceScan(
 
-    val rowtime = tableSource match {
+    val withRowtime = tableSource match {
       case timeSource: DefinedRowtimeAttribute if timeSource.getRowtimeAttribute != null =>
```
Can you open an issue for this? We can discuss this after merging this PR.

wuchong replied:
Sure, I have logged it https://issues.apache.org/jira/browse/FLINK-7446
